src/AppInstallerCommonCore/MSStore.cpp

```diff
@@ -175,11 +249,43 @@ namespace AppInstaller::MSStore
         std::vector<AppInstallItem> installItemVector{ installItem };
         IVectorView<AppInstallItem> installItems = winrt::single_threaded_vector(std::move(installItemVector)).GetView();
 
+        // Check if we need to restart or cancel existing items.
+        if (RestartOrCancelExistingOperationIfNecessary(installItems, installManager, m_productId))
+        {
+            // Wait a bit
+            Sleep(500);
+
+            // Try again
+            installItem = installManager.SearchForUpdatesAsync(
+                m_productId,          // ProductId
+                winrt::hstring(),   // SkuId
+                winrt::hstring(),
+                winrt::hstring(),   // ClientId
+                updateOptions
+            ).get();
+
+            if (!installItem)
+            {
+                return APPINSTALLER_CLI_ERROR_UPDATE_NOT_APPLICABLE;
+            }
+
+            installItemVector.clear();
+            installItemVector.emplace_back(installItem);
+            installItems = winrt::single_threaded_vector(std::move(installItemVector)).GetView();
+        }
+
         return WaitForOperation(installItems, progress);
     }
 
     HRESULT MSStoreOperation::WaitForOperation(IVectorView<AppInstallItem>& installItems, IProgressCallback& progress)
     {
+        auto cancelIfOperationFailed = wil::scope_exit(
+            [&]()
+            {
+                AppInstallManager installManager;
+                installManager.Cancel(m_productId);

```

Comment:
We probably want to try/catch this whole lambda to prevent any errors from potentially terminating the process.